### PR TITLE
test: Add integration tests for Patient and Practitioner FHIR facade

### LINKED ISSUE #3648: Create More Integration Tests for PractitionerProvider and Patient Provider

### PR DESCRIPTION
**Pull request by kavya685**
# Pull Requests Requirements
- [x] The PR title includes a brief description of the work done, including the
      Issue number if applicable.
- [ ] The PR includes a video showing the changes for the work done.
- [x] The PR title follows conventional commit label standards.
- [x] The changes confirm to the OpenElis Global x3
      [Styleguide and Design](https://uwdigi.atlassian.net/wiki/spaces/OG/pages/621346838/OpenELIS+Global+Style+Guide)
      documentation.
- [x] The changes include tests or are validated by existing tests.
- [x] I have read and agree to the Contributing
      [Guidelines](https://openelis-global.org/community/get-involved/) of this
      project.

## Summary
Adds comprehensive integration tests for the `PatientProvider` and 
`PractitionerProvider` FHIR facade endpoints, as requested in #3648.

## Changes
Tests Added (34 Total):
  PatientFacadeTest (19 tests)
  PractitionerFacadeTest (15 tests)

* Implemented ResourceNotFoundException throw/rethrow to ensure a clean 404 Not Found response is returned to the client.

  ---

  Test Results
   * Tests Run: 34
   * Failures: 0
   * Errors: 0
   * Build Status: SUCCESS

## Screenshots
tests (2)

## Related Issue
Closes #3648
[Add a link to the related issue or mention it here if applicable]

## Other

[Add any additional information or notes here]

### CODE REVIEW

File: src/main/java/org/openelisglobal/fhir/providers/PractitionerProvider.java [hunk 1]
```diff
@@ -148,6 +148,9 @@ public MethodOutcome update(@IdParam IdType theId, @ResourceParam Practitioner p
 
             Provider provider = providerService
                     .getProviderByFhirId(UUID.fromString(practitioner.getIdElement().getIdPart()));
+            if (provider == null) {
+                throw new ResourceNotFoundException("Practitioner/" + theId.getIdPart());
```
Comment (Agaba-derrick) [inline src/main/java/org/openelisglobal/fhir/providers/PractitionerProvider.java:152]: why have we added this ?

**Comment by kavya685** [comment]
Hi @pmanko @Isabirye1515, I have submitted the PR [#3650](https://github.com/DIGI-UW/OpenELIS-Global-2/pull/3650) with the requested integration tests. It covers CRUD operations for both Patient and Practitioner, includes persistence verification, handles edge cases, and fixes a small issue in PractitionerProvider where it was returning a 500 error instead of a 404. Looking forward to your feedback!

File: src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java [hunk 2]
```diff
@@ -263,4 +263,293 @@ public void deletePatient_withNonExistentId_shouldReturn404() throws Exception {
...
+        JsonNode json = objectMapper.readTree(response.getContentAsString());
+        assertEquals("Patient", json.get("resourceType").asText());
+        assertEquals(patientUuid, json.get("id").asText());
+        assertTrue("Response should contain a name", json.has("name"));
+        assertTrue("Response should contain gender", json.has("gender"));
+    }
+
+    @Test
+    public void readPatient_femalePatient_shouldReturnFemaleGender() throws Exception {
+        // fixture patient id=2 has gender=F
```
Comment (Agaba-derrick) [inline src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java]: these ccomments are not required, lets look at spliting test responsibilities 

HTTP contract
persistence
mapping

Not all three mixed randomly

File: src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java [hunk 3]
```diff
@@ -263,4 +263,293 @@ public void deletePatient_withNonExistentId_shouldReturn404() throws Exception {
 
         assertEquals(404, response.getStatus());
```
Comment (Agaba-derrick) [inline src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java]: these tests are integration-style in execution
but not dataset-driven in design
and partially rely on hidden DB state assumptions

**Review by Agaba-derrick — CHANGES_REQUESTED** [review]

**Comment by Agaba-derrick** [comment]
lets create  a screen short of these tests passing !
lets avoid test comments, lets  make sure the test is xml dataset driven
A good example is [ MenuServiceTest](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/develop/src/test/java/org/openelisglobal/menu/MenuServiceTest.java)
and the [PatientService Test
](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/develop/src/test/java/org/openelisglobal/patient/service/PatientServiceTest.java)

File: src/main/java/org/openelisglobal/fhir/providers/PractitionerProvider.java (hunk 1 above)
Comment (kavya685) [reply src/main/java/org/openelisglobal/fhir/providers/PractitionerProvider.java:152]: Hi @Agaba-derrick, thank you for the feedback!

Regarding the null check: I added it because the original implementation was throwing a 500 error when an update was attempted on a non-existent ID. This change ensures we return a proper 404 status, which aligns with standard RESTful practices. I'm happy to adjust if there is a specific pattern preferred in this project.

File: src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java (hunk 2 above)
Comment (kavya685) [reply src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java]: You're right—I can definitely clean this up. I will refactor the tests to separate the HTTP contract logic from persistence validation and use more 'dataset-driven' patterns instead of relying on existing state. Could you point me to an existing test class in the repo that best demonstrates the test design pattern you'd like me to follow?

**Comment by kavya685** [comment]
> lets create a screen short of these tests passing ! lets avoid test comments, lets make sure the test is xml dataset driven A good example is [ MenuServiceTest](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/develop/src/test/java/org/openelisglobal/menu/MenuServiceTest.java) and the [PatientService Test ](https://github.com/DIGI-UW/OpenELIS-Global-2/blob/develop/src/test/java/org/openelisglobal/patient/service/PatientServiceTest.java)

Thanks again for the guidance, @Agaba-derrick! I'm going to refactor these to be dataset-driven and remove the comments as requested. I expect to have the updated code and the test results screenshot pushed within the next couple of hours. I'll let you know once those are up for your review.

**Comment by kavya685** [comment]
I have uploaded screenshort in PR description

File: src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java [hunk 4]
```diff
@@ -75,234 +57,37 @@ public void setUp() throws Exception {
     }
 
     private MockHttpServletRequest buildRequest(String method, String pathInfo) {
```
Comment (Isabirye1515) [inline src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java]: This can be removed completely, already defined in BaseWebContextSensitiveTest

File: src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java [hunk 5]
```diff
@@ -75,234 +57,37 @@ public void setUp() throws Exception {
...
         assertEquals("Patient", json.get("resourceType").asText());
-        assertEquals(patientUuid, json.get("id").asText());
-        assertEquals("male", json.get("gender").asText());
-    }
-
-    @Test
-    public void createPatient_withBirthDate_shouldPersistCorrectly() throws Exception {
-        clearDataBase();
-
-        MockHttpServletRequest request = buildRequest("POST", "/Patient");
```
Comment (Isabirye1515) [inline src/test/java/org/openelisglobal/fhir/PatientFacadeTest.java:141]: Why r u completely deleting this, we have to just see afew adjustments

File: src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java [hunk 6]
```diff
@@ -396,98 +288,29 @@ public void updatePractitioner_shouldPersistNewGivenName() throws Exception {
     }
 
     @Test
-    public void deletePractitioner_thenRecordShouldBeInactive() throws Exception {
```
Comment (Isabirye1515) [inline src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java]: can u adjust the code without these deletions please

**Comment by Isabirye1515** [comment]
Thanks @kavya685 , am sorry for late review, but our IG has some items that are mandatory , can u include tests for these items ie Address for patient is 1..1, and others please include these tests, thanks 

**OpenELIS Implementation Guide**: https://digi-uw.github.io/openelis-global2-fhir-ig/StructureDefinition-openelis-patient.html

File: src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java [hunk 7]
```diff
@@ -29,215 +28,295 @@
...
+    public void createPractitioner_shouldReturn201() throws Exception {
+        prepareCleanSlate();
+        MockHttpServletRequest request = buildFhirRequest("POST", "/Practitioner");
         String practitionerJson = """
                 {
                   "resourceType": "Practitioner",
                   "active": true,
-                  "name": [
-                    {
-                      "use": "official",
```
Comment (Isabirye1515) [inline src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java:133]: Please leave a clear reason for these deletions

**Review by Isabirye1515 — COMMENTED** [review]
Can you try running**mvn spotless:apply** before pushing, cause having many unnecesary changes makes the work untidy, make a few necessary changes

File: src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java (hunk 7 above)
Comment (kavya685) [reply src/test/java/org/openelisglobal/fhir/PractitionerFacadeTest.java:133]: To split the combined test into focused single-assertion tests

**Review by mozzy11 — COMMENTED** [review]
Really valuable contribution, @kavya685 — facade test coverage was explicitly requested in #3648 and 34 passing integration tests plus the clean 404 handling is exactly what we needed. One coordination point: #3874 is rewriting PractitionerProvider search to query the local database and also adds a PractitionerFacadeTest, so your Practitioner-side tests and provider change will collide with it. Suggested path: we merge the practitioner search rework first, then you rebase — the PatientFacadeTest and facade-patient.xml fixture should carry over untouched, and the Practitioner tests can be adapted to exercise the new local-search behavior (arguably making them even more valuable). I'll approve the CI run now so we can confirm the suite is green on current develop.

**Review by mozzy11 — COMMENTED** [review]
Re-reviewed the June 18 rework against current develop — this has come a long way and, from my read, addresses the substance of @Agaba-derrick's review: tests are now dataset-driven (`facade-patient.xml` extended with SUBJECT/ST/GUID identity types), responsibilities are split into focused single-assertion tests (HTTP contract vs mapping vs persistence), the inline comments are gone, and the base-class helpers (`buildFhirRequest`, `executeDataSetWithStateManagement`, `cleanRowsInCurrentConnection`) are used instead of local copies.

I also verified the diff base matches current develop exactly (PatientFacadeTest 708151e60, PractitionerProvider 14d9082d2), and the `PractitionerProvider.update()` null-guard is still needed — develop's update() still NPEs to a 500 on a nonexistent id, while delete() already 404s. The IG mandatory-element coverage @Isabirye1515 asked about is partially there (address asserted on read/create); worth confirming that thread is settled.

@Agaba-derrick your CHANGES_REQUESTED is the blocking state — please take another look; if the rework covers your points, a re-review would unblock this. One coordination note (flagged 2026-07-17): #3720/#3874 rewrite PractitionerProvider search, so Practitioner-side merges should be sequenced with those.
